test/expr/closure/closures.swift

```diff
@@ -246,9 +246,9 @@ class ExplicitSelfRequiredTest {
     
     // If we already have a capture list, self should be added to the list
     let y = 1
-    doStuff { [y] in method() } // expected-warning {{capture 'y' was never used}} expected-error {{call to method 'method' in closure requires explicit use of 'self' to make capture semantics explicit}} expected-note{{capture 'self' explicitly to enable implicit 'self' in this closure}} {{16-16=self, }} expected-note{{reference 'self.' explicitly}} {{22-22=self.}}
+    doStuff { [y] in method() } // expected-error {{call to method 'method' in closure requires explicit use of 'self' to make capture semantics explicit}} expected-note{{capture 'self' explicitly to enable implicit 'self' in this closure}} {{16-16=self, }} expected-note{{reference 'self.' explicitly}} {{22-22=self.}}
     doStuff { [ // expected-note{{capture 'self' explicitly to enable implicit 'self' in this closure}} {{16-16=self, }}
-        y // expected-warning {{capture 'y' was never used}}
+        y
```

Comment:
The change in these warnings suggests the pass changes are over-broad.

Reply:
Having a bit of trouble understanding why this specific warning isn't getting emitted in the test case -- when I use this toolchain to compile a standalone with the same shape as this example, I'm seeing the expected warning:

```swift
func doStuff(_ fn : @escaping () -> Int) {}

class Test {
    let y = 10
    func method() {}

    func test() {
      doVoidStuff { [y] in
        method()
      }
    }
}
```

```
/Users/cal/swift-samples/weak.swift:10:22: warning: capture 'y' was never used
      doVoidStuff { [y] in
                     ^
```

Plus the other errors and notes:

```
/Users/cal/swift-samples/weak.swift:11:9: error: call to method 'method' in closure requires explicit use of 'self' to make capture semantics explicit
        method()
        ^
/Users/cal/swift-samples/weak.swift:11:9: note: reference 'self.' explicitly
        method()
        ^
        self.
/Users/cal/swift-samples/weak.swift:10:19: note: capture 'self' explicitly to enable implicit 'self' in this closure
      doVoidStuff { [y] in
                  ^
                     self, 
```

Reply:
Honestly, the VarDeclUsageChecker is a giant ad-hoc analysis. You may have just shifted the AST enough to evade its pattern matching. Unfortunately, that means some debugging is in order!

Reply:
Aha, I figured out what is going on here one I reduced a repro case:

```swift
func doVoidStuff(_ fn : @escaping () -> ()) {}
func doStuff(_ fn : @escaping () -> Int) {}

class ExplicitSelfRequiredTest {
  var x = 42
  
  func test() {
    doVoidStuff({ [y = self] in x += 1 })
    
    // error: explicit use of 'self' is required when 'self' is optional, to make control flow explicit
    // note: reference 'self?.' explicitly
    doVoidStuff({ [weak self] in x += 1 })
  }
  
  func method() -> Int { }
}
```

The second invalid `doVoidStuff` call is emitting an error early enough that `VarDeclUsageChecker`'s `sawError` flag is being set to true, which causes it to not run any analysis on the `func test() { }` method body. Once you fix that error (e.g. by inserting the `self?.` recommended by the fixit), then the `capture 'y' was never used` warning is emitted on the following run.

So I think moving the invalid `weak self` closures to their own method in the `closures.swift` test should fix this.

Reply:
These tests all pass now, including the expected `capture 'y' was never used` warnings 👍🏻